server/src/main/java/org/elasticsearch/action/admin/indices/settings/get/GetSettingsResponse.java

```diff
@@ -71,5 +111,135 @@ public void writeTo(StreamOutput out) throws IOException {
+                indexToSettings.put(currentIndexName, Settings.fromXContent(parser));
+                break;
+            case "defaults":
+                indexToDefaultSettings.put(currentIndexName, Settings.fromXContent(parser));
+                break;
+            default:
+                throw new IllegalStateException("Unknown field name for settings field!  Should be \"settings\" or \"default\", but " +
+                    "was: "+settingsFieldName);
+
+        }
+    }
+
+    public static GetSettingsResponse fromXContent(XContentParser parser) throws IOException {
+        HashMap<String, Settings> indexToSettings = new HashMap<>();
+        HashMap<String, Settings> indexToDefaultSettings = new HashMap<>();
+
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+        parser.nextToken();
+
+        while (!parser.isClosed() && parser.currentToken() == XContentParser.Token.FIELD_NAME) {
+            String indexName = parser.currentName();
+            parser.nextToken(); //go to per-index settings object
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+            parser.nextToken();
+
+            parseSettingsField(parser, indexName, indexToSettings, indexToDefaultSettings);  //this will get one of settings or defaults
+            parser.nextToken();
+
+            if (parser.currentToken() == XContentParser.Token.FIELD_NAME) {
+                //we have both settings and default settings, let's grab the second one
+                parseSettingsField(parser, indexName, indexToSettings, indexToDefaultSettings);
+                parser.nextToken();
+            }
+
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.currentToken(), parser::getTokenLocation);
+            parser.nextToken(); //we should now be positioned at the beginning of the next index's settings, or at the end
+        }
+
+        ImmutableOpenMap<String, Settings> settingsMap = ImmutableOpenMap.<String, Settings>builder().putAll(indexToSettings).build();
+        ImmutableOpenMap<String, Settings> defaultSettingsMap =
+            ImmutableOpenMap.<String, Settings>builder().putAll(indexToDefaultSettings).build();
+
+
+        return new GetSettingsResponse(settingsMap, defaultSettingsMap);
+    }
+
+    @Override
+    public String toString() {
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            XContentBuilder builder = new XContentBuilder(JsonXContent.jsonXContent, baos);
+            toXContent(builder, ToXContent.EMPTY_PARAMS, false);
+            return Strings.toString(builder);
+        } catch (IOException e) {
+            throw new IllegalStateException(e); //should not be possible here
+        }
```

Comment:
why not using `Strings#toString(ToXContent)` here?

Reply:
I see why but I am not sure that the new flag that you pass to toXContent helps that much, given that it's only used in the `toString` method? I think that we can do without it and keep things simple. what do you think?

Reply:
I'd recommend keeping it -- the reason is that toString() method is called and displayed for both the expected and actual objects during tests.  If we "lie" in toString(), the output in the case of a test failure may appear to be identical when in fact it is not.  I ran into this during testing and it is precisely why I made it this way (I had it as a plain toXContent before).

Reply:
I didn't realize that the toString could lie without this added flag. I don't understand though completely why it's useful. Could you post an example of a situation where it helps please? Like what you have seen while testing.

Reply:
gotcha, thanks for the explanation! I'd prefer to simplify things a bit here. After all this is a corner case that happens only in our unit tests, an index will always have at least a couple of settings like `number_of_shards`, `creation_date` etc. I would rather move the `if` to the transport action, or even better, remove this optimization given that it buys us nothing (provided no tests fail after such change :) )

Reply:
By the way, having no settings to show could happen in practice when one provides a settings_filter that doesn't match any settings. I think it would be more correct to still return the index in that case, with an empty set of settings. Problem is that at the moment the API doesn't accept settings_filter, and this is a bug that we should fix separately. Working on these little client PRs tends to uncover a lot of stuff :)

Reply:
ok on settings_filter param that is not supported here, I will open a followup PR, I think that such param should not have been exposed anywhere after all. But we do allow to filter through the names parameter instead, so it is possible that we have no settings to return for some index. Let's keep this behaviour for backwards compatibility reasons but let's change it as a follow-up in master and get rid of this toXContent variant and toString impl?